phosphor buttons: Enable Delayed/Fast Power Off

sbteeks opened the pull request:
The current Power Button behavior is to power On or Off host with a
short press. The power Off behaviour on FSP-based platforms is
different, and is follows:
Initial 4 seconds: do nothing
Delayed Power Off Initiated: With this, 10 seconds countdown begins.
If Button is released during this interval and not pressed again,
normal shutdown - equivalent to 'obmcutil poweroff' - starts and
the host goes to Off state. If the Button is released and pressed
again during this interval OR button is held down beyond this interval,
Fast Power Off is initiated with 10 seconds countdown, and the host is
shutdown - equivalent of 'obmcutil chassisoff'.
This commit changes the behavior on p10bmc platforms similar to
FSP-based platforms.

The following tests were performed using simulator signals and
power button on rainier system:
1. Power On host (push/release button)
   Verified: host in 'Running' state
2. Power Off (Release Button within 4s)
   Verified: No change
3. Power Off (Release Button somewhere between 5-9s)
   Verified: host in 'Off' state
4. Power Off (Release Button after 6s, and press/release again)
   Verified: host in 'Off' state
5. Power Off (Release Button after 6s, press and hold it down 15s)
   Verified: host in 'Off' state
6. Power Off (Release after 30s)
   Verified: host in 'Off' state

Signed-off-by: Shantappa Teekappanavar <sbteeks@yahoo.com>
Change-Id: Ie52d1fd598971776c085e7ba9ce9e5264bd9d508

sbteeks commented on src/power_button.cpp:
```diff
@@ -16,27 +16,313 @@
+    }
+    catch (const sdbusplus::exception::exception& e)
+    {
+        log<level::DEBUG>(
```
Level should be ERR, will fix this along with other comments if any...

spinler commented on inc/button_interface.hpp:
```diff
@@ -52,6 +52,7 @@ class ButtonIface
             buttonIface->handleEvent(es, fd, revents);
             return 0;
         }
+        return 0;
```
there's a gerrit commit at https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-buttons/+/51183/3/inc/button_interface.hpp#55 that also removes the return at line 53

sbteeks replied:
Looks like a bit major work to support power operations with a single button. Going to be interesting to merge as we are diverging from upstream. 

I'll take out the return statement at line 53. Thanks for pointing the gerrit link...

spinler commented on src/power_button.cpp:
```diff
@@ -16,27 +16,313 @@
+    std::stringstream ss;
+
+    // System is On, initiate the power Off process
+    if (powerOpState == PowerOpState::buttonNotPressed)
```
I was thinking you'd do all of this in the button handler, as this class already emits signals on when buttons are pressed and released.

sbteeks replied:
Was thinking about that, but then I thought button handler is meant for handling the signal to power on/off purpose only. Will move this to button handler.

spinler replied:
Thanks.  That gerrit commit I pointed to also added some new functionality, and also just to button handler so this would match that.  That commit will also merge imminently so you could probably even pull that down and rebase yours on that one so it is closer to upstream, and possibly even upstreamable itself.

sbteeks replied:
Posted changes. This is without the new functionality in gerrit.

sbteeks committed:
phosphor buttons: Enable Delayed/Fast Power Off

The current Power Button behavior is to power On or Off host with a
short press. The power Off behaviour on FSP-based platforms is
different, and is follows:
Initial 4 seconds: do nothing
Delayed Power Off Initiated: With this, 10 seconds countdown begins.
If Button is released during this interval and not pressed again,
normal shutdown - equivalent to 'obmcutil poweroff' - starts and
the host goes to Off state. If the Button is released and pressed
again during this interval OR button is held down beyond this interval,
Fast Power Off is initiated with 10 seconds countdown, and the host is
shutdown - equivalent of 'obmcutil chassisoff'.
This commit changes the behavior on p10bmc platforms similar to
FSP-based platforms.

The following tests were performed using simulator signals and
power button on rainier system:
1. Power On host (push/release button)
   Verified: host in 'Running' state
2. Power Off (Release Button within 4s)
   Verified: No change
3. Power Off (Release Button somewhere between 5-9s)
   Verified: host in 'Off' state
4. Power Off (Release Button after 6s, and press/release again)
   Verified: host in 'Off' state
5. Power Off (Release Button after 6s, press and hold it down 15s)
   Verified: host in 'Off' state
6. Power Off (Release after 30s)
   Verified: host in 'Off' state

Signed-off-by: Shantappa Teekappanavar <sbteeks@yahoo.com>
Change-Id: If8e4955ee748edb50df458dd64eb805b8406ea8a

spinler commented on src/button_handler_main.cpp:
```diff
@@ -1,15 +1,41 @@
 #include "button_handler.hpp"
+#include "common.hpp"
+
+#include <phosphor-logging/elog-errors.hpp>
```
to get the log<>() function, you just need phosphor-logging/log.hpp

spinler commented:
Looks pretty good to me!
Any need for any more field traces?

I'd like to get at least one other reviewer before approving.

sbteeks commented:
Removed some of the DEBUG traces, but the rest I would like to keep just in case if we end-up debugging an issue in the future - but can be deleted once we have the SRC codes directed to Op-panel.
